src/rust/engine/remote_provider/src/lib.rs

```diff
@@ -0,0 +1,128 @@
+// Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+#![deny(warnings)]
+// Enable all clippy lints except for many of the pedantic ones. It's a shame this needs to be copied and pasted across crates, but there doesn't appear to be a way to include inner attributes from a common source.
+#![deny(
+  clippy::all,
+  clippy::default_trait_access,
+  clippy::expl_impl_clone_on_copy,
+  clippy::if_not_else,
+  clippy::needless_continue,
+  clippy::unseparated_literal_suffix,
+  clippy::used_underscore_binding
+)]
+// It is often more clear to show that nothing is being moved.
+#![allow(clippy::match_ref_pats)]
+// Subjective style.
+#![allow(
+  clippy::len_without_is_empty,
+  clippy::redundant_field_names,
+  clippy::too_many_arguments
+)]
+// Default isn't as big a deal as people seem to think it is.
+#![allow(clippy::new_without_default, clippy::new_ret_no_self)]
+// Arc<Mutex> can be more clear than needing to grok Orderings:
+#![allow(clippy::mutex_atomic)]
+
+use std::sync::Arc;
+
+// Re-export these so that consumers don't have to know about the exact arrangement of underlying
+// crates.
+pub use remote_provider_traits::{
+  ActionCacheProvider, ByteStoreProvider, LoadDestination, RemoteCacheProviderOptions,
+  RemoteOptions,
+};
+
+const REAPI_ADDRESS_SCHEMAS: [&str; 4] = ["grpc://", "grpcs://", "http://", "https://"];
+
+// TODO(#19902): a unified view of choosing a provider would be nice
```

Comment:
These two functions are very similar, but I haven't tried to merge them as part of this work. There's two closely-related points in #19902 for follow-up ("tweaking the internal Rust details of how the remote providers are built to have less duplication", "improving how the options are fed through").